src/ray/raylet/node_manager.cc

```diff
@@ -2897,25 +2897,52 @@ MemoryUsageRefreshCallback NodeManager::CreateMemoryUsageRefreshCallback() {
             << "worker pid: " << high_memory_eviction_target_->GetProcess().GetId()
             << "task: " << high_memory_eviction_target_->GetAssignedTaskId();
       } else {
-        auto newest_task_start_time = std::chrono::high_resolution_clock::time_point();
-        std::shared_ptr<WorkerInterface> latest_worker = nullptr;
-        for (const auto &worker : worker_pool_.GetAllRegisteredWorkers()) {
-          if (worker->GetAssignedTaskTime() > newest_task_start_time) {
-            newest_task_start_time = worker->GetAssignedTaskTime();
-            latest_worker = worker;
+        auto workers = worker_pool_.GetAllRegisteredWorkers();
+        if (!workers.empty()) {
+          std::sort(workers.begin(),
+                    workers.end(),
+                    [](std::shared_ptr<WorkerInterface> const &left,
+                       std::shared_ptr<WorkerInterface> const &right) -> bool {
+                      return left->GetAssignedTaskTime() > right->GetAssignedTaskTime();
+                    });
+          std::shared_ptr<WorkerInterface> latest_worker = workers.front();
+          std::ostringstream top_n_latest_workers_info;
+          int64_t top_n = 0;
+
+          const static int64_t max_to_print = 10;
+          for (auto &worker : workers) {
+            auto pid = worker->GetProcess().GetId();
+            auto used_memory = this->memory_monitor_->GetProcessMemoryBytes(pid);
+            auto task_assign_time =
+                std::chrono::system_clock::to_time_t(worker->GetAssignedTaskTime());
+            top_n_latest_workers_info
+                << "Worker " << top_n << " task assigned time "
+                << std::put_time(std::gmtime(&task_assign_time), "%c %Z")
+                << " memory used " << used_memory << " task spec "
+                << worker->GetAssignedTask().GetTaskSpecification().DebugString() << "\n";
+            top_n += 1;
+            if (top_n >= max_to_print) {
+              break;
+            }
           }
-        }
-        if (latest_worker != nullptr) {
-          RAY_LOG(INFO) << "Killing worker with the newest started task "
-                        << "to free up memory. "
-                        << "worker pid: " << latest_worker->GetProcess().GetId()
-                        << "task: " << latest_worker->GetAssignedTaskId();
+          RAY_LOG_EVERY_N_OR(ERROR, 10, INFO)
+              << "Killing worker with the latest task assigned time "
+              << "to free up memory. Task to be killed:"
+              << latest_worker->GetAssignedTaskId() << ". Top 10 worker details:\n"
```

Comment:
Also mention the current memory usage (used / total)

Reply:
Should we direct them to the raylet log? i.e.

[2022-08-17 02:34:49,122 I 99866 99901] (raylet) memory_monitor.cc:71: Node memory usage above threshold, used: 25104216064, total: 35738255360, usage fraction: 0.
702447, threshold: 0.7

[2022-08-17 02:34:49,260 E 99866 99901] (raylet) node_manager.cc:2926: [0] Killing worker with the latest task assigned time to free up memory. Find details of workers killed via `ray list workers --detail`. Task to be killed:fffffffffffffffff14084600f04d906c4ae596701000000. Top 10 worker details:

Reply:
Agree with @rkooo567, we should include the memory usage here so it's all in one place and make sure it's human-readable.

Also, can we make the instructions about what to do next more specific? For example, populate the `ray list workers` command with the ID of the worker that we're going to kill? Ideally, we should also include short instructions for how to avoid the problem, such as provisioning more memory and what Ray configuration to set if they want to increase the memory threshold.